mindmeld/components/dialogue.py

```diff
@@ -689,6 +691,201 @@ def _get_dialogue_handler(self, dialogue_state):
         return handler
 
 
+class AutoEntityFilling(DialogueFlow):
+    """A special dialogue flow sublcass to implement Automatic Entitiy (Slot) Filling
+    (AEF) that allows developers to prompt users for completing the missing
+    requirements for entity slots.
+
+    Attributes:
+        app (Application): The application that initializes this flow.        
+    """
+
+    logger = mod_logger.getChild("AutoEntityFilling")
+    """Class logger."""
+
+    def __init__(self, name, entrance_handler, app, **kwargs):
+        super().__init__(name, entrance_handler, app)
+
+        self.app = app
+        self.entrance_handler = entrance_handler
+
+    def _set_target_state(self):
+        """Set target dialogue state to current flow"""
+        responder.params.target_dialogue_state = self.flow_state
+
+    def _exit_flow(self):
+        """Exits this flow and clears the related parameter for re-usability"""
+        if 'slot_not_prompted' in responder.frame:
+            del responder.frame['slot_not_prompted']
+        responder.exit_flow()
+
+    def _extract_query_features(self, text):
+        """ Extracts Query object from the user input and converts it into
+        appropriate format for entity extraction.
+
+        Args:
+            text (Request.text): Text in the query
+
+        Returns:
+            Formatted Payload (tuple(query (Query Object), list of entities, entity index))
+        """
+        resource_loader = self.app.app_manager.nlp.resource_loader
+
+        """ Extracting the Query object associated with the user input here.
+
+        process_markup args:
+            text, query factory, query options (optional dict) {language, time_zone and time_stamp}
+
+        returns: 
+            tuple(raw_text (str), Query object, list of entities)
+        """
+        query = process_markup(text, resource_loader.query_factory, {})[1]
+
+
+        """ 
+        payload format for entity feature extractors: 
+            tuple(query (Query Object), list of entities, entity index)
+
+        For slot filling, the user query will consist of the prompted entity. 
+        Hence `entity = [query]` and `entity_index = 0`.
+
+        """
+        formatted_payload = (query, [query], 0)
+        return formatted_payload
+
+    def _validate(self, text, entity_type):
```

Comment:
It might be worth separating out the duckling, ulist, and gazetteer validations into three separate public methods that the developer could leverage if they decide to define their own validation function and want to use one of those as part of it. We can then call those methods from here as well